src/tests/testCacheManager.cc

```diff
@@ -66,3 +75,86 @@ testCacheManager::testRegister()
     CPPUNIT_ASSERT_EQUAL(1,(int)sentry->flags);
 }
 
+void
+testCacheManager::testParseUrl()
+{
+    auto *mgr = static_cast<CacheManagerInternals *>(CacheManager::GetInstance());
+    CPPUNIT_ASSERT(mgr != nullptr);
+
+    std::vector<AnyP::ProtocolType> validSchemes = {
+        AnyP::PROTO_CACHE_OBJECT,
+        AnyP::PROTO_HTTP,
+        AnyP::PROTO_HTTPS,
+        AnyP::PROTO_FTP
+    };
+
+    AnyP::Uri mgrUrl;
+    mgrUrl.host("localhost");
+    mgrUrl.port(3128);
+
+    const std::vector<const char *> validPathActions = {
+        // "", // technically valid, but not supported
+        "/",
+        "/menu",
+        // "/squid-internal-mgr", // technically valid, but not supported
+        "/squid-internal-mgr/",
+        "/squid-internal-mgr/menu"
+    };
+
+    const std::vector<const char *> validParams = {
+        "",
+        "?",
+        "?&",
+        "?&&&&&&&&&&&&",
+        "?foo=bar",
+        "?0123456789=bar",
+        "?foo=bar&",
+        "?foo=bar&&&&",
+        "?&foo=bar",
+        "?&&&&foo=bar",
+        "?&foo=bar&",
+        "?&&&&foo=bar&&&&",
+        "?foo=?_weird?~`:[]stuff&bar=okay&&&&&&"
+    };
+
+    const std::vector<const char *> validFragments = {
+        "",
+        "#frag"
+    };
+
+    unsigned caseNumber = 0;
+    unsigned success = 0;
+    for (const auto &scheme : validSchemes) {
+        mgrUrl.setScheme(scheme);
+
+        for (const auto *action : validPathActions) {
+
+            // all schemes except cache_object require "/squid-internal-mgr" path prefix
+            if (scheme != AnyP::PROTO_CACHE_OBJECT && strlen(action) <= 19)
+                continue;
+
+            for (const auto *param : validParams) {
+
+                for (const auto *frag : validFragments) {
+                    try {
+                        ++caseNumber;
+                        std::cerr << caseNumber << std::endl;
+
+                        SBuf bits;
+                        bits.append(action);
+                        bits.append(param);
+                        bits.append(frag);
+                        mgrUrl.path(bits);
+
+                        (void)mgr->ParseUrl(mgrUrl);
+                        ++success;
+                    } catch (...) {
+                        std::cerr << std::endl << "FAIL: " << mgrUrl << std::endl
+                                  << CurrentException << std::endl;
+                    }
+                    CPPUNIT_ASSERT_EQUAL(caseNumber, success);
```

Comment:
I do not understand the intent behind this line. I could understand the desire to fail on the first error or the desire to fail after all possible errors/cases, but this does neither AFAICT. Please adjust the code to either explain the odd choice or, better, to match one of those natural intents.

Reply:
It exists for cppunit tracking of tests run. It does halt on the first error - by means of any parser thrown exceptions skipping the ++success code statement.

Reply:
> It does halt on the first error

Please adjust the code to do that explicitly rather than to rely on the numerical equality between apples (the current case ID) and oranges (the number of cases that did not throw in the past), triggering reader doubts (at best). AFAICT, we should move the CppUnit assert (or equivalent) to the catch clause. If done right, this move would also improve failure diagnostic, possibly significantly.

Reply:
Done (I think) in 351e7ba56.